pkg/kubelet/eviction/helpers.go

```diff
@@ -1011,6 +1014,11 @@ func evictionMessage(resourceToReclaim v1.ResourceName, pod *v1.Pod, stats stats
 		for _, container := range pod.Spec.Containers {
 			if container.Name == containerStats.Name {
 				requests := container.Resources.Requests[resourceToReclaim]
+				if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+					if _, cs, ok := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name); ok {
+						requests = cs.ResourcesAllocated[resourceToReclaim]
```

Comment:
This seems like it could break eviction. I'd like to see test updates to go along with this.

Reply:
We should have test coverage for this case if we are changing the behaviour when the feature flag is on, and I don't see any corresponding test changes.

I don't understand why we override the requests with the ResourcesAllocated when VPA is enabled. Below,

```
if usage != nil && usage.Cmp(requests) > 0
```

So this changes the eviction thresholds for a pod that's scaled depending on the status ResourcesAllocated rather than container spec Requests. I can't find anything in the KEP describing a behaviour change on evictions when a pod is undergoing VPA, so trying to understand what the intended result is here.

Rereading [the KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes) it seems that if these don't match, we are in the midst of a resize, but we don't check the resize status before we use ResourcesAllocated here. I would expect that we would still use the old field, since according to the spec that is the desired state of the world, whether or not we have reached it yet:

> Pod.Spec.Containers[i].Resources becomes purely a declaration, denoting the desired state of Pod resources,

Am I misreading this?

Reply:
Trying to jog my own memory so please bear with me. IIRC, with Requests becoming mutable (desired value), relying on it would affect the pod's ranking as an eviction candidate in unpredictable/undesirable ways. We should evict based on what a pod is has been actually granted (i.e. ResourcesAllocated which is the value kubelet agreed to give the pod upon successful admit - regardless of whether that admit was initial or if it was from a successful resize)
Does that clarify anything?

Reply:
@vinaykul I agree that we should evict based on actual resources allocated so this change is correct.

I agree with @ehashman that as part of this feature going further we need to make a small update to `helpers_test` in eviction module for the tests that are `TestOrderedByXYZ` for the resource types that are adjustable.  It could be as simple as changing the `newResourceRequirements` helper method to set ResourcesAllocated to match the request values.

Reply:
I had to add a new test to tap the evictionMessage function. None of the current tests appear to trigger it. Fix will be in the next commit.

Reply:
New test was added for to resolve this conversation.
https://github.com/kubernetes/kubernetes/pull/102884/files#diff-f9c89695d37a7bb54bdde095ec05213c567af15170c18614550ecc573c78c545R2131